src/lamp_py/ingestion/convert_gtfs_rt.py

```diff
@@ -397,33 +385,76 @@ def sync_with_s3(self, local_path: str) -> bool:
-        new_table_path = os.path.join(
-            local_folder,
-            "new_table.parquet",
+        part_col = self.detail.partition_column
+        table = hash_gtfs_rt_table(table)
+        out_ds = pd.dataset(table)
+
+        if self.sync_with_s3(local_path) is True:
+            hash_gtfs_rt_parquet(local_path)
+            out_ds = pd.dataset(
+                [
+                    pd.dataset(table),
+                    pd.dataset(local_path),
+                ]
+            )
+
+        partitions = pc.unique(
+            out_ds.to_table(columns=[part_col]).column(part_col)
         )
-        pq.write_table(table, new_table_path)
 
-        join_table_path = os.path.join(
-            local_folder,
-            "join_table.parquet",
+        with tempfile.TemporaryDirectory() as temp_dir:
+            tmp_pq = os.path.join(temp_dir, "temp.parquet")
+            with pq.ParquetWriter(tmp_pq, schema=out_ds.schema) as writer:
+                for part in partitions:
+                    part_filter = pc.field(part_col) == part
+                    table = pl.from_arrow(out_ds.to_table(filter=part_filter))
+                    writer.write_table(
+                        table.sort(
+                            by=["feed_timestamp"],
+                        )
+                        .unique(subset=GTFS_RT_HASH_COL, keep="first")
+                        .to_arrow()
+                        .cast(out_ds.schema)
+                    )
+
+            os.replace(tmp_pq, local_path)
+
+    def upload_local_pq(self, local_path: str) -> None:
+        """
+        upload local parquet file to S3
+        """
+        part_col = self.detail.partition_column
+        upload_path = local_path.replace(
+            self.tmp_folder, os.environ["SPRINGBOARD_BUCKET"]
         )
-        join_ds = pd.dataset((local_path, new_table_path)).sort_by(
-            self.detail.table_sort_order
+        ds = pd.dataset(local_path)
+        write_columns = ds.schema.names
+        write_columns.remove(GTFS_RT_HASH_COL)
+        write_schema = ds.schema.remove(
+            ds.schema.get_field_index(GTFS_RT_HASH_COL)
         )
-        with pq.ParquetWriter(join_table_path, schema=table.schema) as writer:
-            for batch in join_ds.to_batches(batch_size=self.pq_batch_size):
-                writer.write_batch(batch)
+        with tempfile.TemporaryDirectory() as temp_dir:
+            tmp_pq = os.path.join(temp_dir, "temp.parquet")
+            with pq.ParquetWriter(tmp_pq, schema=write_schema) as writer:
+                for part in pc.unique(
+                    ds.to_table(columns=[part_col]).column(part_col)
+                ):
+                    writer.write_table(
+                        ds.to_table(
+                            columns=write_columns,
+                            filter=(pc.field(part_col) == part),
+                        ).sort_by(self.detail.table_sort_order)
+                    )
```

Comment:
drop hash column from dataset before uploading to S3, the hash column adds a decent amount to the parquet file size and external users should not need this column